Error 500 after a provider is updated

Reproducible in released 4.2.1:

Steps:

- Log into oxtrust with admin user+password. Set oxtrust acr to `passport_social`
- Add & configure 1 or more providers. For simplicity you can use github, twitter...
- Tail the passport.log in a terminal
- In oxtrust modify some property of one of the providers, say, the display name
- Back to the terminal, in less than one min, the following appears:

```
2020-09-26T12:32:25.094Z [ERROR]    TypeError: Strategy is not a constructor
2020-09-26T12:32:25.094Z [DEBUG]    TypeError: Strategy is not a constructor
    at setupStrategy (/opt/gluu/node/passport/server/providers.js:111:22)
    at forEach (/opt/gluu/node/passport/node_modules/ramda/src/forEach.js:45:5)
    at /opt/gluu/node/passport/node_modules/ramda/src/internal/_checkForMethod.js:22:72
    at Object.f2 [as forEach] (/opt/gluu/node/passport/node_modules/ramda/src/internal/_curry2.js:29:14)
    at Object.setup (/opt/gluu/node/passport/server/providers.js:239:7)
    at reconfigure (/opt/gluu/node/passport/server/app.js:98:13)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
2020-09-26T12:32:25.094Z [WARN]     An attempt to get configuration data will be tried again soon
```

- Logout, or in a different browser, visit oxtrust. The configured providers appear in the right-hand side pane. 
- Click on any of them, you will get a 500 error page
- Passport.log shows

```
2020-09-26T12:32:31.433Z [VERBOSE]  Issuing token
2020-09-26T12:32:31.454Z [INFO]     ::1 - GET /passport/token HTTP/1.1 200 201 - 6.586 ms
2020-09-26T12:32:31.477Z [VERBOSE]  Validating token
2020-09-26T12:32:31.479Z [VERBOSE]  Authenticating request against twitter
2020-09-26T12:32:31.479Z [ERROR]    Unknown Error: Error: Unknown authentication strategy "twitter"
2020-09-26T12:32:31.480Z [ERROR]    Error: Unknown authentication strategy "twitter"
    at attempt (/opt/gluu/node/passport/node_modules/passport/lib/middleware/authenticate.js:190:39)
    at authenticate (/opt/gluu/node/passport/node_modules/passport/lib/middleware/authenticate.js:367:7)
    at authenticateRequest (/opt/gluu/node/passport/server/routes.js:108:77)
    at Layer.handle [as handle_request] (/opt/gluu/node/passport/node_modules/express/lib/router/layer.js:95:5)
    at next (/opt/gluu/node/passport/node_modules/express/lib/router/route.js:137:13)
    at validateToken (/opt/gluu/node/passport/server/routes.js:136:5)
    at Layer.handle [as handle_request] (/opt/gluu/node/passport/node_modules/express/lib/router/layer.js:95:5)
    at next (/opt/gluu/node/passport/node_modules/express/lib/router/route.js:137:13)
    at validateProvider (/opt/gluu/node/passport/server/routes.js:100:5)
    at Layer.handle [as handle_request] (/opt/gluu/node/passport/node_modules/express/lib/router/layer.js:95:5)
    at next (/opt/gluu/node/passport/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/opt/gluu/node/passport/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/opt/gluu/node/passport/node_modules/express/lib/router/layer.js:95:5)
    at /opt/gluu/node/passport/node_modules/express/lib/router/index.js:281:22
    at param (/opt/gluu/node/passport/node_modules/express/lib/router/index.js:354:14)
    at param (/opt/gluu/node/passport/node_modules/express/lib/router/index.js:365:14)
    at param (/opt/gluu/node/passport/node_modules/express/lib/router/index.js:365:14)
    at Function.process_params (/opt/gluu/node/passport/node_modules/express/lib/router/index.js:410:3)
    at next (/opt/gluu/node/passport/node_modules/express/lib/router/index.js:275:10)
    at Function.handle (/opt/gluu/node/passport/node_modules/express/lib/router/index.js:174:3)
    at router (/opt/gluu/node/passport/node_modules/express/lib/router/index.js:47:12)
    at Layer.handle [as handle_request] (/opt/gluu/node/passport/node_modules/express/lib/router/layer.js:95:5)
2020-09-26T12:32:31.487Z [INFO]     ::1 - GET /passport/auth/twitter/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqd3QiOiIyYWNhYjRlYS05ZDAwLTQ1M2ItYjhlOC02MWI1OWJmZmU3YTkiLCJpYXQiOjE2MDExMjM1NTEsImV4cCI6MTYwMTEyMzY3MX0.jWk-bNVA4djBsWH05tVjoM4mjSIkKM_9MR-PN8Aa024 HTTP/1.1 500 148 - 9.859 ms
```

Happily the issue resolves automatically because reconfiguration takes place every minute, however in that 1 min window, all login requests using external identity providers will face 500.

Problem is [here](https://github.com/GluuFederation/gluu-passport/blob/version_4.2.1/server/providers.js#L73) and [here](https://github.com/GluuFederation/gluu-passport/blob/version_4.2.1/server/providers.js#L80)

The array `passportStrategies` is supposed to store objects with keys `id` and `strategy`, but now we have `Strategy`. Basically a typo. See this [comment](https://github.com/GluuFederation/gluu-passport/blob/version_4.2.1/server/providers.js#L12) as well.

In 4.2 it was correct: [here](https://github.com/GluuFederation/gluu-passport/blob/version_4.2.1/server/providers.js#L12) and [here](https://github.com/GluuFederation/gluu-passport/blob/version_4.2.0/server/providers.js#L86).

--- a/server/providers.js
+++ b/server/providers.js
@@ -9,7 +9,7 @@ const cacheProvider = require('./cache-provider')
 
 var prevConfigHash = 0
 
-// These are the (node) strategies loaded so far: [{id: "...", strategy: ...}, ... ]
+// These are the (node) strategies loaded so far: [{id: "...", Strategy: ...}, ... ]
 var passportStrategies = []
 
 function applyMapping (profile, provider) {
@@ -70,7 +70,7 @@ function setupStrategy (prv) {
 
   // if module is not found, load it
   if (Strategy) {
-    Strategy = Strategy.strategy
+    Strategy = Strategy.Strategy
   } else {
     logger.log2('info', `Loading node module ${moduleId}`)
     Strategy = require(moduleId)

--- a/test/providers.test.js
+++ b/test/providers.test.js
@@ -1,0 +1,46 @@
+const chai = require('chai')
+const rewire = require('rewire')
+const providers = rewire('../server/providers.js')
+const testConfig = require('../config/test')
+
+const assert = chai.assert
+
+describe('providers setupStrategy', () => {
+  const passportStrategies = providers.__get__('passportStrategies')
+
+  it('passport strategies array should be empty first time', () => {
+    assert.deepEqual(passportStrategies, [])
+  })
+
+  it('new strategy should be added in passport strategies', () => {
+    const setupStrategy = providers.__get__('setupStrategy')
+    const testProvider = {
+      ...testConfig.passportConfigAuthorizedResponse.providers[0],
+      verifyCallbackArity: 0
+    }
+
+    setupStrategy(testProvider)
+
+    assert.lengthOf(passportStrategies, 1)
+    assert.isFunction(
+      passportStrategies[0].Strategy,
+      'Strategy is not a function!'
+    )
+  })
+
+  it('existing loaded strategy should be found and load again', () => {
+    const setupStrategy = providers.__get__('setupStrategy')
+    const testProvider = {
+      ...testConfig.passportConfigAuthorizedResponse.providers[0],
+      verifyCallbackArity: 0
+    }
+
+    setupStrategy(testProvider)
+
+    assert.lengthOf(passportStrategies, 1)
+    assert.isFunction(
+      passportStrategies[0].Strategy,
+      'Strategy is not a function!'
+    )
+  })
+})

--- a/test/providers.test.js
+++ b/test/providers.test.js
@@ -7,18 +7,17 @@ const assert = chai.assert
 
 describe('providers setupStrategy', () => {
   const passportStrategies = providers.__get__('passportStrategies')
+  const setupStrategy = providers.__get__('setupStrategy')
+  const testProvider = {
+    ...testConfig.passportConfigAuthorizedResponse.providers[0],
+    verifyCallbackArity: 0
+  }
 
   it('passport strategies array should be empty first time', () => {
     assert.deepEqual(passportStrategies, [])
   })
 
   it('new strategy should be added in passport strategies', () => {
-    const setupStrategy = providers.__get__('setupStrategy')
-    const testProvider = {
-      ...testConfig.passportConfigAuthorizedResponse.providers[0],
-      verifyCallbackArity: 0
-    }
-
     setupStrategy(testProvider)
 
     assert.lengthOf(passportStrategies, 1)
@@ -29,12 +28,6 @@ describe('providers setupStrategy', () => {
   })
 
   it('existing loaded strategy should be found and load again', () => {
-    const setupStrategy = providers.__get__('setupStrategy')
-    const testProvider = {
-      ...testConfig.passportConfigAuthorizedResponse.providers[0],
-      verifyCallbackArity: 0
-    }
-
     setupStrategy(testProvider)
 
     assert.lengthOf(passportStrategies, 1)

